larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+    std::vector<UintFloatPair> sliceIndexProbabilityPairs;
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+    {
+        const float nuProbability(sliceFeaturesVector.at(sliceIndex).GetNeutrinoProbability(m_supportVectorMachine));
+
+        if (nuProbability < m_minProbability)
+        {
+            this->SelectPfos(crSliceHypotheses.at(sliceIndex), selectedPfos);
+            continue;
+        }
+
+        sliceIndexProbabilityPairs.push_back(UintFloatPair(sliceIndex, nuProbability));
+    }
+
+    // Sort the slices by probability
+    std::sort(sliceIndexProbabilityPairs.begin(), sliceIndexProbabilityPairs.end(), [] (const UintFloatPair &a, const UintFloatPair &b) 
+    {
+        return (a.second > b.second);
+    });
+
+    // Select the first m_maxNeutrinos as neutrinos, and the rest as cosmic
+    unsigned int nNuSlices(0);
+    for (const UintFloatPair &slice : sliceIndexProbabilityPairs)
+    {
+        if (nNuSlices < m_maxNeutrinos)
+        {
+            this->SelectPfos(nuSliceHypotheses.at(slice.first), selectedPfos);
+            nNuSlices++;
+            continue;
+        }
+        
+        this->SelectPfos(crSliceHypotheses.at(slice.first), selectedPfos);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::SelectPfos(const PfoList &pfos, PfoList &selectedPfos) const
+{
+    selectedPfos.insert(selectedPfos.end(), pfos.begin(), pfos.end());
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+// TODO think about how to make this function cleaner when features are more established
+NeutrinoIdTool::SliceFeatures::SliceFeatures(const PfoList &nuPfos, const PfoList &crPfos, const NeutrinoIdTool *const pTool) : 
+    m_isAvailable(false), m_pTool(pTool)
+{
+    const ParticleFlowObject *const pNeutrino(this->GetNeutrino(nuPfos));
+    const CartesianVector nuVertex(LArPfoHelper::GetVertex(pNeutrino)->GetPosition());
+    const PfoList &nuFinalStates(pNeutrino->GetDaughterPfoList());
+
+    // Neutrino features
+    CartesianVector nuWeightedDirTotal(0, 0, 0);
+    unsigned int nuNHitsUsedTotal(0);
+    unsigned int nuNHitsTotal(0);
+    CartesianPointVector nuAllSpacePoints;
+    for (const ParticleFlowObject *const pPfo : nuFinalStates)
+    {
+        CartesianPointVector spacePoints;
+        this->GetSpacePoints(pPfo, spacePoints);
+
+        nuAllSpacePoints.insert(nuAllSpacePoints.end(), spacePoints.begin(), spacePoints.end());
+        nuNHitsTotal += spacePoints.size();
+
+        if (spacePoints.size() < 5) continue;
```

Comment:
A few hardcoded numbers to (ultimately) replace via e.g. xml-configured parameters.

Reply:
I was wary about making these configurable parameters, because if they are changed then the SVM training could be thrown off. Maybe it is best for them to still be configurable nonetheless?